scheduler/core/components/optimizer/timeline.py

```diff
@@ -25,11 +26,109 @@ class Timeline:
+        """
+        return self.get_available_intervals()[0]
+
+    def add(self, iobs: int, time_slots: int, interval: int) -> datetime:
+        """
+        Add an observation index to the first open position (-1) in the given interval
+        Returns the time of this position
+        """
+        # interval = self.get_earliest_available_interval()
+        # Get first non-zero slot in given interval
+        i_first = np.where(self.time_slots[interval] == EMPTY)[0][0]
+
+        # Set values of time_slots to the observation index
+        self.time_slots[interval[i_first:i_first + time_slots]] = iobs
+
+        # Clock time for the starting index
+        start = self.start + interval[i_first] * self.time_slot_length
+
+        return start
+
+    def get_observation_order(self) -> List[Tuple[int]]:
+        """
+        Get the observation idx and position for all the schedule observations in order
+        Return
+        -------
+
+        orders: List of a 3-dimensional tuple: observation.idx, initial position
+                and last position in the plan.
+        """
+
+        schedule = self.time_slots
+        obs_comparator = self.time_slots[0]
+        start = 0
+
+        order = []
+        for position, obs_idx in enumerate(schedule):
+            if obs_idx != obs_comparator:
+                order.append((obs_comparator, start, position - 1))
+                start = position
+                obs_comparator = obs_idx
+            elif position == len(schedule) - 1:
+                order.append((obs_comparator, start, position))
+
+        return order
+
+    def __str__(self):
+        """Print the result of get_observation_order, an experiment in the use of __str__"""
+
+        obs_order = self.get_observation_order()
+
+        return f"{obs_order}"
+
+    def print(self, obsids) -> NoReturn:
+        """Print the obsids and times associated with the timeline"""
+
+        delta = timedelta(milliseconds=500) # for rounding to the nearest second
+
+        obs_order = self.get_observation_order()
+
+        for idx, i_start, i_end in obs_order:
+            obsid = obsids[idx]
+            if idx > -1:
+                # Convert time slot indices to UT
+                t_start = self.start.astimezone(tz=timezone.utc) + i_start * self.time_slot_length + delta
+                # t_end = self.start.astimezone(tz=timezone.utc) + i_end * self.time_slot_length + delta
+                t_end = t_start + (i_end - i_start) * self.time_slot_length
+                
+                print("{:5d} {:5d} {:5d}   {:20} {:}  {:}".format(idx, i_start, i_end, obsid,
+                                                               # t_start.isoformat(), t_end.isoformat())
+                                                             t_start.strftime('%Y-%m-%d %H:%M:%S'),
+                                                             t_end.strftime('%Y-%m-%d %H:%M:%S'))
+                      )
```

Comment:
`str.format` is considered deprecated. Combine into an f-string instead.

Replace with:
```python
print(f'{idx:5d} {i_start:5d} {i_end:5d}   {obs_id:20} {t_start.strftime("%Y-%m-%d %H:%M:%S")} '
                      f'{t_end.strftime("%Y-%m-%d %H:%M:%S")}')
```

I broke it into two lines to avoid going over the line length limit, but the strings automatically concatenate without need for `+` since they are between parentheses.